File: app/src/test/java/software/credible/abercrombiefitchkata/domain/PromotionUnitTest.java
package software.credible.abercrombiefitchkata.domain;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.ArrayList;
import java.util.Arrays;

import software.credible.abercrombiefitchkata.dto.ButtonDto;
import software.credible.abercrombiefitchkata.dto.PromotionDto;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

/**
 * Unit tests for promotions.
 */
@RunWith(JUnit4.class)
public class PromotionUnitTest {

    @Test
    public void itTranslatesFromPromotionDtoToPromotionWithoutLosingData() {

        PromotionDto promotionDto = new PromotionDto();
        promotionDto.setTitle("Title");
        promotionDto.setImageUrl("ImageUrl");
        promotionDto.setDescription("Description");
        promotionDto.setFooter("Footer");

        ButtonDto buttonDto = new ButtonDto();
        buttonDto.setTitle("ButtonTitle");
        buttonDto.setTargetUrl("ButtonTargetUrl");
        promotionDto.setButtons(new ArrayList<>(Arrays.asList(buttonDto)));


        Promotion promotion = Promotion.fromDto(promotionDto);


        assertEquals("Title", promotion.getTitle());
        assertEquals("ImageUrl", promotion.getImageUrl());
        assertEquals("Description", promotion.getDescription());
        assertEquals("Footer", promotion.getFooter());
        assertEquals(1, promotion.getButtons().size());
        Button button = promotion.getButtons().get(0);
        assertEquals("ButtonTitle", button.getTitle());
        assertEquals("ButtonTargetUrl", button.getTargetUrl());
    }

    @Test
    public void itGeneratesAnIdWhenTranslatingFromPromotionDtoToPromotion() {
        PromotionDto dto = new PromotionDto();
        Promotion promotion = Promotion.fromDto(dto);
        assertNotNull(promotion.getId());
    }

    @Test
    public void itHandlesTranslatingFromPromotionDtoToPromotionGracefullyWhenThereAreNoButtonsSet() {
        PromotionDto dto = new PromotionDto();
        Promotion promotion = Promotion.fromDto(dto);
        assertNotNull(promotion.getId());
    }

}

File: app/src/test/java/software/credible/abercrombiefitchkata/domain/PromotionUnitTest.kt
package software.credible.abercrombiefitchkata.domain

import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

import java.util.ArrayList
import java.util.Arrays

import software.credible.abercrombiefitchkata.dto.ButtonDto
import software.credible.abercrombiefitchkata.dto.PromotionDto

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull

/**
 * Unit tests for promotions.
 */
@RunWith(JUnit4::class)
class PromotionUnitTest {

    @Test
    fun itTranslatesFromPromotionDtoToPromotionWithoutLosingData() {

        val promotionDto = PromotionDto()
        promotionDto.title = "Title"
        promotionDto.imageUrl = "ImageUrl"
        promotionDto.description = "Description"
        promotionDto.footer = "Footer"

        val buttonDto = ButtonDto()
        buttonDto.title = "ButtonTitle"
        buttonDto.targetUrl = "ButtonTargetUrl"
        promotionDto.buttons = ArrayList(Arrays.asList(buttonDto))


        val promotion = Promotion.fromDto(promotionDto)

        assertEquals("Title", promotion.title)
        assertEquals("ImageUrl", promotion.imageUrl)
        assertEquals("Description", promotion.description)
        assertEquals("Footer", promotion.footer)
        assertEquals(1, promotion.buttons?.size)
        val button = promotion.buttons?.get(0)
        assertEquals("ButtonTitle", button!!.title)
        assertEquals("ButtonTargetUrl", button.targetUrl)
    }

    @Test
    fun itGeneratesAnIdWhenTranslatingFromPromotionDtoToPromotion() {
        val dto = PromotionDto()
        val promotion = Promotion.fromDto(dto)
        assertNotNull(promotion.id)
    }

    @Test
    fun itHandlesTranslatingFromPromotionDtoToPromotionGracefullyWhenThereAreNoButtonsSet() {
        val dto = PromotionDto()
        val promotion = Promotion.fromDto(dto)
        assertNotNull(promotion.id)
    }

}
